cmd/kubeadm/app/util/apiclient/idempotency.go

```diff
@@ -51,6 +54,38 @@ func CreateOrUpdateConfigMap(client clientset.Interface, cm *v1.ConfigMap) error
 	return nil
 }
 
+// CreateOrMutateConfigMap tries to create the ConfigMap provided as cm. If the resource exists already, the latest version will be fetched from
+// the cluster and mutator callback will be called on it, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the ConfigMap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func CreateOrMutateConfigMap(client clientset.Interface, cm *v1.ConfigMap, mutator ConfigMapMutator) error {
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "unable to create ConfigMap")
+		}
+		return MutateConfigMap(client, metav1.ObjectMeta{Namespace: cm.ObjectMeta.Namespace, Name: cm.ObjectMeta.Name}, mutator)
+	}
+	return nil
+}
+
+// MutateConfigMap takes a ConfigMap Object Meta (namespace and name), retrieves the resource from the server and tries to mutate it
+// by calling to the mutator callback, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the ConfigMap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func MutateConfigMap(client clientset.Interface, meta metav1.ObjectMeta, mutator ConfigMapMutator) error {
+	return clientsetretry.RetryOnConflict(clientsetretry.DefaultRetry, func() error {
+		configMap, err := client.CoreV1().ConfigMaps(meta.Namespace).Get(meta.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if err = mutator(configMap); err != nil {
+			return errors.Wrap(err, "unable to mutate ConfigMap")
+		}
+		_, err = client.CoreV1().ConfigMaps(configMap.ObjectMeta.Namespace).Update(configMap)
```

Comment:
How are you ensuring multiple concurrent writers aren't overwriting each other?

Reply:
Only a Patch would be safe in this scenario.

Reply:
* [client1] Fetch `kubeadm-config` (v1)
* [client2] Fetch `kubeadm-config` (v1)
* [client1] Mutate `kubeadm-config` locally (v1)
* [client2] Mutate `kubeadm-config` locally (v1)
* [client1] Update `kubeadm-config` locally (v2)
* [client2] Update `kubeadm-config` locally (v1) - CONFLICT
* [client2] Fetch `kubeadm-config`  (v2)
* [client2] Mutate `kubeadm-config` locally (v2)
* [client2] Update `kubeadm-config` locally (v3)

`Patch` vs `Update` does not give more or less safety in this case, the conflict will happen in both cases, `Patch` is just handier to only provide a subset of properties to be altered. At least this is my understanding, please tell me if I'm understanding something in a wrong way.

Reply:
Update doesn't care though it's blind and dump, there is no conflict, it's just a raw put. 
```
result = &v1.ConfigMap{}
        err = c.client.Put().
                Namespace(c.ns).
                Resource("configmaps").
                Name(configMap.Name).
                Body(configMap).
                Do().
                Into(result)
```
There is no conflict it's just a straight overwrite.

Reply:
Patch would indeed give you more safety b/c order would not matter.

Reply:
I can ensure `Update` does indeed return a conflict if the version of the resource that you are updating is not the latest stored in `etcd`.

Reply:
So long as resource version is decoded from the `get` and passed again on the update you're right.  Sorry, it's been too long since I thought about this.